src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -491,13 +492,26 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
             CancellationTokenSource cts;
             bool disposeCts;
             bool hasTimeout = _timeout != s_infiniteTimeout;
+            TimeoutState timeoutState = null; // Lazy initialization to avoid unnecessary alocation when timeout is not set
+            Timer timer = null;
             if (hasTimeout || cancellationToken.CanBeCanceled)
             {
                 disposeCts = true;
                 cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
                 if (hasTimeout)
                 {
-                    cts.CancelAfter(_timeout);
+                    timeoutState = new TimeoutState(cts);
+                    timer = new Timer(s =>
+                    {
+                        var state = (TimeoutState)s;
+                        state.IsFired = true;
+                        try
+                        {
+                            state.CancellationSource.Cancel();
+                        }
+                        catch (ObjectDisposedException) // ObjectDisposedException is expected when the timer fires just after the CTS has been disposed
+                        { }
+                    }, timeoutState, _timeout, Threading.Timeout.InfiniteTimeSpan);
```

Comment:
It's unfortunate we're adding another allocation.  We could explore ways to avoid it, e.g. using a type like:
```C#
private sealed class LinkedTimedCancellationTokenSource : CancellationTokenSource
{
    private readonly Timer _timer;
    private readonly CancellationTokenRegistration _reg1;
    private readonly CancellationTokenRegistration _reg2;

    public volatile bool TimerFired;

    public LinkedTimedCancellationTokenSource(CancellationToken token1, CancellationToken token2, TimeSpan timeout)
    {
        Action<object> cancelCallback = s => ((LinkedTimedCancellationTokenSource)s).Cancel();
        _reg1 = token1.UnsafeRegister(cancelCallback);
        _reg2 = token2.UnsafeRegister(cancelCallback);
        _timer = new Timer(s =>
        {
            var thisRef = (LinkedTimedCancellationTokenSource)s;
            thisRef.TimerFired = true;
            try { thisRef.Cancel(); } catch (ObjectDisposedException) { }
        }, this, timeout, Timeout.InfiniteTimeSpan);
    }

    protected override void Dispose(bool disposing)
    {
        _timer.Dispose();
        _reg2.Dispose();
        _reg1.Dispose();
        base.Dispose(disposing);
    }
}
```
and then constructing one of these when a timeout is set, e.g.
```C#
if (hasTimeout || cancellationToken.CanBeCanceled)
{
    disposeCts = true;
    cts = hasTimeout ?
        LinkedTimedCancellationTokenSource(cancellationToken, _pendingRequestsCts.Token, timeout) :
        CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
}
```
That then also implicitly handles the rooting issue, avoids the need to pass around the additional Timer (since it's encapsulated in the CTS), etc.

I'm not sure if it's worth it, but we already allocate a bunch of objects per request, and one of the things we want to look at in this release is driving that down, so increasing it is going in the wrong direction, especially when we end up paying it on all requests in order to handle an exceptional case.